write-fonts/src/font_builder.rs

```diff
@@ -93,35 +127,99 @@ impl<'a> FontBuilder<'a> {
+            } else {
+                (1, 0, tag)
+            }
+        });
+
+        ordered_tags
+    }
+
     /// Assemble all the tables into a binary font file with a [Table Directory].
     ///
     /// [Table Directory]: https://learn.microsoft.com/en-us/typography/opentype/spec/otff#table-directory
+    /// [Calculating Checksums]: https://learn.microsoft.com/en-us/typography/opentype/spec/otff#calculating-checksums
     pub fn build(&mut self) -> Vec<u8> {
         let header_len = std::mem::size_of::<u32>() // sfnt
             + std::mem::size_of::<u16>() * 4 // num_tables to range_shift
             + self.tables.len() * TABLE_RECORD_LEN;
 
+        // note this is the order of the tables themselves, not the records in the table directory
+        // which are sorted by tag so they can be binary searched
+        let table_order = self.ordered_tags();
+
         let mut position = header_len as u32;
-        let table_records: Vec<_> = self
-            .tables
-            .iter_mut()
-            .map(|(tag, data)| {
-                let offset = position;
-                let length = data.len() as u32;
-                position += length;
-                let (checksum, padding) = checksum_and_padding(data);
-                position += padding;
-                TableRecord::new(*tag, checksum, offset, length)
-            })
-            .collect();
+        let mut checksums = Vec::new();
+        let head_tag = Tag::new(b"head");
+
+        let mut table_records = Vec::new();
+        for tag in table_order.iter() {
+            // safe to unwrap as ordered_tags() guarantees that all keys exist
+            let data = self.tables.get_mut(tag).unwrap();
+            let offset = position;
+            let length = data.len() as u32;
+            position += length;
+            if *tag == head_tag {
+                // The head table checksum is computed with the checksum field set to 0.
+                // Equivalent to Python's `data[:8] + b"\0\0\0\0" + data[12:]`
+                let head = data.to_mut();
+                head[8..12].copy_from_slice(&[0, 0, 0, 0]);
+            }
+            let (checksum, padding) = checksum_and_padding(data);
+            checksums.push(checksum);
+            position += padding;
+            table_records.push(TableRecord::new(*tag, checksum, offset, length));
+        }
+        table_records.sort_unstable_by_key(|record| record.tag);
 
         let directory = TableDirectory::from_table_records(table_records);
 
         let mut writer = TableWriter::default();
         directory.write_into(&mut writer);
         let mut data = writer.into_data().bytes;
-        for table in self.tables.values() {
-            data.extend_from_slice(table);
+        checksums.push(read_fonts::tables::compute_checksum(&data));
+
+        let checksum = checksums.into_iter().fold(0u32, u32::wrapping_add);
+        let checksum_adjustment = 0xB1B0_AFBAu32.wrapping_sub(checksum); // BiboAfba!
```

Comment:
this deserves a more expository comment :)

Reply:
fair enough :) I wanted to keep the magic aura 🤣